### PR TITLE
only sync groups if they change

### PR DESCRIPTION
**Pull request by dvoet**
Ticket: https://broadworkbench.atlassian.net/browse/POL-60

the work on POL-60 I think will trigger more group syncing so this is an optimization to limit how many group sync messages we send.

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://broadinstitute.github.io/dsp-appsec-security-risk-assessment/) and attached the result to the JIRA ticket

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 1]
```diff
@@ -428,29 +436,34 @@ class ResourceService(
...
   }
 
-    private def fireGroupUpdateNotification(groupId: WorkbenchGroupIdentity, samRequestContext: SamRequestContext): Future[Unit] =
-    cloudExtensions.onGroupUpdate(Seq(groupId), samRequestContext) recover {
-      case t: Throwable =>
-        logger.error(s"error calling cloudExtensions.onGroupUpdate for $groupId", t)
-        throw t
-    }
+  private def fireGroupUpdateNotification(groupId: WorkbenchGroupIdentity, samRequestContext: SamRequestContext): IO[Unit] =
+    IO.fromFuture(IO(cloudExtensions.onGroupUpdate(Seq(groupId), samRequestContext))).runAsync {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:440]: Note that this is run async which is in line with how it worked before. This is why the tests have timeouts. I also added some timings (see GoogleExtensions) so we can get a sense of how long this actually takes to see if we can make it synchronous. I view this as an integrity issue. If the instance goes down during this process the change has been committed but messages may be dropped. Perhaps a better solution is to have this be a single message then have the group sync machinery do this.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 2]
```diff
@@ -292,18 +290,25 @@ class ResourceService(
...
     * Overwrites an existing policy's membership (keyed by resourceType/resourceId/policyName) if it exists
-    * @param resourceType
-    * @param policyName
-    * @param resource
+    * @param policyId
     * @param membersList
     * @return
     */
-  def overwritePolicyMembers(resourceType: ResourceType, policyName: AccessPolicyName, resource: FullyQualifiedResourceId, membersList: Set[WorkbenchEmail], samRequestContext: SamRequestContext): IO[Unit] =
+  def overwritePolicyMembers(policyId: FullyQualifiedPolicyId, membersList: Set[WorkbenchEmail], samRequestContext: SamRequestContext): IO[Unit] =
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:297]: `overwritePolicyMembers` did not trigger a sync at all 😱 !

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala [hunk 3]
```diff
@@ -186,6 +186,7 @@ class GoogleExtensions(
    */
   override def onGroupUpdate(groupIdentities: Seq[WorkbenchGroupIdentity], samRequestContext: SamRequestContext): Future[Unit] = {
     for {
+      start <- clock.instantNow
```
Comment (s-rubenstein) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:189]: What is the goal of the clock timing here? Just useful log messaging?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala (hunk 2 above)
Comment (s-rubenstein) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:297]: Oh wow. Yeah, that's not great, good catch.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 4]
```diff
@@ -321,15 +326,18 @@ class ResourceService(
     val workbenchSubjects = policy.emailsToSubjects.values.flatten.toSet
     accessPolicyDAO.loadPolicy(policyIdentity, samRequestContext).flatMap {
       case None => createPolicy(policyIdentity, workbenchSubjects, generateGroupEmail(), policy.roles, policy.actions, policy.descendantPermissions, samRequestContext)
-      case Some(accessPolicy) =>
-        for {
-          result <- accessPolicyDAO.overwritePolicy(AccessPolicy(policyIdentity, workbenchSubjects, accessPolicy.email, policy.roles, policy.actions, policy.descendantPermissions, accessPolicy.public), samRequestContext)
-          _ <- IO.fromFuture(IO(fireGroupUpdateNotification(policyIdentity, samRequestContext))).runAsync {
-            case Left(regrets) => IO(logger.error(s"failure calling fireGroupUpdateNotification on $policyIdentity", regrets))
```
Comment (s-rubenstein) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:328]: How come we don't need the error logging anymore?

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceServiceSpec.scala [hunk 5]
```diff
@@ -843,6 +958,90 @@ class ResourceServiceSpec extends AnyFlatSpec with Matchers with ScalaFutures wi
...
+    val mockCloudExtensions: CloudExtensions = mock[CloudExtensions](RETURNS_SMART_NULLS)
+    val mockDirectoryDAO: DirectoryDAO = mock[DirectoryDAO](RETURNS_SMART_NULLS)
+    val resourceService = new ResourceService(Map.empty,
+      mock[PolicyEvaluatorService](RETURNS_SMART_NULLS),
+      mock[AccessPolicyDAO](RETURNS_SMART_NULLS),
+      mockDirectoryDAO,
+      mockCloudExtensions,
+      "")
+
+    val policyId = FullyQualifiedPolicyId(FullyQualifiedResourceId(defaultResourceType.name, ResourceId("testR")), AccessPolicyName("testA"))
```
Comment (s-rubenstein) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceServiceSpec.scala:971]: TOL: I noticed `testR`, `testA`, and `testU` are used a lot across these tests, maybe it makes sense to make a variable or name them higher up? Unless we're testing on responses I'm not too worried, but I could see that popping up.

**Review by s-rubenstein — COMMENTED** [review]
Looks fairly good to me, a few things to talk through, I think?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala [hunk 6]
```diff
@@ -128,25 +128,19 @@ class ManagedGroupService(
...
     val resourceAndPolicyName =
       FullyQualifiedPolicyId(FullyQualifiedResourceId(ManagedGroupService.managedGroupTypeName, resourceId), policyName)
-    accessPolicyDAO.loadPolicy(resourceAndPolicyName, samRequestContext).flatMap {
-      case Some(policy) => {
-        val updatedPolicy = AccessPolicyMembership(emails, policy.actions, policy.roles, Option(policy.descendantPermissions))
-        resourceService.overwritePolicy(managedGroupType, resourceAndPolicyName.accessPolicyName, resourceAndPolicyName.resource, updatedPolicy, samRequestContext)
-      }
-      case None => throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"Group or policy could not be found: $resourceAndPolicyName"))
-    }
+    resourceService.overwritePolicyMembers(resourceAndPolicyName, emails, samRequestContext)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala:134]: this was simplified by switching from `overwritePolicy` to `overwritePolicyMembers`

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala (hunk 4 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala:328]: it moved, see line 441

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala (hunk 3 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:189]: https://github.com/broadinstitute/sam/pull/535/files#r672506030

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceServiceSpec.scala (hunk 5 above)
Comment (dvoet) [reply src/test/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceServiceSpec.scala:971]: the values are irrelevant for the tests themselves and I err on the side of keeping tests self contained

**Review by s-rubenstein — APPROVED** [review]
LGTM

**Review by marctalbott — APPROVED** [review]

**Pull request merged**